iOS/DuckDuckGo/MaliciousSiteProtection/Services/MaliciousSiteProtectionDatasetsFetcher.swift

```diff
@@ -60,36 +73,60 @@ final class MaliciousSiteProtectionDatasetsFetcher: MaliciousSiteProtectionDatas
         userPreferencesManager: MaliciousSiteProtectionPreferencesProvider,
         dateProvider: @escaping () -> Date = Date.init,
         backgroundTaskScheduler: BGTaskScheduling = BGTaskScheduler.shared,
-        application: BackgroundRefreshCapable = UIApplication.shared
+        application: BackgroundRefreshCapable = UIApplication.shared,
+        preferencesScheduler: AnySchedulerOf<DispatchQueue> = DispatchQueue.main.eraseToAnyScheduler()
     ) {
         self.updateManager = updateManager
         self.featureFlagger = featureFlagger
         self.userPreferencesManager = userPreferencesManager
         self.dateProvider = dateProvider
         self.backgroundTaskScheduler = backgroundTaskScheduler
         self.application = application
+        self.preferencesScheduler = preferencesScheduler
     }
 }
 
 // MARK: - Public
 
-extension MaliciousSiteProtectionDatasetsFetcher {
+extension MaliciousSiteProtectionDatasetsFetcher: MaliciousSiteProtectionDatasetsFetching {
 
-    func startFetching() {
-        guard canFetchDatasets else { return }
+    @MainActor
+    @discardableResult
+    func startFetching() -> Task<Void, Error> {
+        guard canFetchDatasets else { return Task {} }
+
+        var updateTasksToReturn: [Task<Void, Error>] = []
 
         Logger.MaliciousSiteProtection.datasetsFetcher.debug("Feature is On and Enabled in App Settings")
 
         // If hashPrefix Sets need to be updated fetch them
         if shouldUpdateHashPrefixSets {
             Logger.MaliciousSiteProtection.datasetsFetcher.debug("Downloading HashPrefixSets")
-            _ = updateManager.updateData(datasetType: .hashPrefixSet)
+            let task = updateManager.updateData(datasetType: .hashPrefixSet)
+            inFlyUpdateTasks[.hashPrefixSet] = task
+            Task {
+                try? await task.value
+                inFlyUpdateTasks[.hashPrefixSet] = nil
```

Comment:
Personal opinion: considering the juggling needed here to make sure asynchronous code actually is reliable, I find it hard to believe we should be adopting it (async/await) widely, as simple queue here would just do the trick in a way that would be far more clear to reason about.